feat(VIcon): add svg support

asvae opened the pull request:
## TODO
- [x] Check VIcon props to work consistently for both svg and font icons.
- [x] Add props support for components.
- [x] Add Font Awesome as preset.

**Finalization**
- [x] Add tests
- [x] Add config docs
- [x] Add faSvg preset docs

## Description
<!--- Describe your changes in detail -->
Allows to add svg icons to icon set. And override existing ones.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
1) Allows user to use his own svg icon (or even img) instead of font icons.
2) Adds vue-font-awesome support. It's useful if you do not want to include whole font.

## How Has This Been Tested?
* added unit tests to `VIcon.spec.js`
* visually (comparison table)
![](http://i64.tinypic.com/15ek4y.png)

## Markup:
<!--- Paste markup for testing your change --->
<details>

Add font awesome icons:
```bash
yarn add @fortawesome/fontawesome-svg-core @fortawesome/vue-fontawesome @fortawesome/free-solid-svg-icons -D
```

In `index.html`:
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Welcome to Vuetify</title>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link href='https://fonts.googleapis.com/css?family=Roboto:100,300,400,500,700,900|Material+Icons' rel="stylesheet" type="text/css">
    <link href="https://use.fontawesome.com/releases/v5.2.0/css/all.css" rel="stylesheet">
  </head>
  <body>
    <div id="app"></div>
    <script src="/dev/main.js"></script>
  </body>
</html>
```

In `index.js`:
```js
import Vue from 'vue'
import App from './App'
import Boilerplate from './Boilerplate'
import Vuetify from 'vuetify'
import VueRouter from 'vue-router'
import router from './router'
import '@mdi/font/css/materialdesignicons.css'

import SvgIcon from './SvgIcon'
import { library } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { fas } from '@fortawesome/free-solid-svg-icons'

// Register component globally
Vue.component('font-awesome-icon', FontAwesomeIcon)
// Include needed icons.
library.add(fas)

Vue.config.performance = true

Vue.use(Vuetify, {
  iconfont: 'faSvg',
  icons: {
    delete: {
      component: SvgIcon,
      props: {
        role: 'icon'
      }
    },
    customSvg: {
      component: SvgIcon,
      props: {
        role: 'icon'
      }
    },
    emptyComponentIcon: {
      component: {render: (h) => h('div', {class: 'custom-component'})}
    },
    faSvg: {
      component: 'font-awesome-icon',
      props: {
        icon: 'air-freshener'
      }
    }
  }
})


Vue.use(VueRouter)

Vue.component(Boilerplate.name, Boilerplate)

const vm = new Vue({
  data: () => ({ isLoaded: document.readyState === 'complete' }),
  render (h) {
    return this.isLoaded ? h(App) : undefined
  },
  router
}).$mount('#app')

// Prevent layout jump while waiting for styles
vm.isLoaded || window.addEventListener('load', () => {
  vm.isLoaded = true
})
```

`SvgIcon.vue`
```vue
<template>
  <svg
    xmlns="http://www.w3.org/2000/svg"
    class="svg-icon"
    viewBox="0 0 528.9 528.9"
    :role="role"
  >
    <path d="M328.88 89.13l107.6 107.58-272.35 272.34L56.6 361.46 328.88 89.13zm189.23-25.95L470.13 15.2a47.61 47.61 0 0 0-67.26 0l-45.96 45.96 107.6 107.59 53.6-53.61a36.68 36.68 0 0 0 0-51.96zM.3 512.69c-1.96 8.81 6 16.7 14.81 14.56l119.9-29.06-107.54-107.6L.3 512.7z"/>
  </svg>
</template>

<script>
export default {
  name: 'SvgIcon',
  props: ['role'],
}
</script>

<style>
.svg-icon {
  fill: currentColor;
}
</style>
```

`Playground.vue`
```vue
<template>
  <v-app>
    <div style="background: #cccccc; overflow: auto">
      <table class="test-table">
        <tr>
          <th>Description</th><th>Material icon</th><th>FA icon</th><th>FA SVG icon</th><th>SVG component</th><th>Passes</th>
        </tr>
        <tr>
          <td>Keeps size</td>
          <td v-for="icon in icons" :key="icon">
            <v-icon>{{ icon }}</v-icon>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Simple
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-icon>{{ icon }}</v-icon>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Colored
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-icon color="success">{{ icon }}</v-icon>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Dark
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-icon dark>{{ icon }}</v-icon>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Disabled
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-icon disabled>{{ icon }}</v-icon>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Size presets
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-icon small>{{ icon }}</v-icon>
            <v-icon medium>{{ icon }}</v-icon>
            <v-icon large>{{ icon }}</v-icon>
            <v-icon x-large>{{ icon }}</v-icon>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Exact size
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-icon size="55px">{{ icon }}</v-icon>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Left/Right
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-btn>
              <v-icon left>{{ icon }}</v-icon>
              Cancel
              <v-icon right>{{ icon }}</v-icon>
            </v-btn>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Light
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-card dark>
              <v-icon light>{{ icon }}</v-icon>
            </v-card>
          </td>
          <td>
            +
          </td>
        </tr>
        <tr>
          <td>
            Works in container
          </td>
          <td v-for="icon in icons" :key="icon">
            <v-card>
              <v-icon>{{ icon }}</v-icon>
            </v-card>
          </td>
          <td>
            +
          </td>
        </tr>
      </table>

      <div>
        SVG icon in component.
        <v-chip close>
          Close icon is svg
        </v-chip>
      </div>

      <div>
        Comparing fa svg and font icons <font-awesome-icon icon="air-freshener"/><i class="fa fa-air-freshener"/>
      </div>

      <div>
        Assortment of fa svg icons if theme is used
        <v-icon>$vuetify.icons.sort</v-icon>
        <v-icon>$vuetify.icons.expand</v-icon>
        <v-icon>$vuetify.icons.menu</v-icon>
      </div>
    </div>
  </v-app>
</template>

<script>
export default {
  data: () => ({}),
  computed: {
    icons: ()  => {
      return [
        'edit',
        'fa fa-air-freshener',
        '$vuetify.icons.faSvg',
        '$vuetify.icons.delete',
      ]
    }
  }
}
</script>

<style>
.test-table {
  border-collapse: collapse;
}
.test-table td, .test-table th {
  padding: 3px;
  box-sizing: border-box;
  border: grey solid 1px;
}
</style>
```
</details>

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.

## Related issues:
* #4340 [Feature Request] Component for Flexible SVG icons

asvae commented:
Current progress:

![](https://i.imgur.com/4ldvGgJ.png)

ekoeryanto commented:
can't wait this feature 😍

DmytroDiachuk commented:
I think,  it is  good solution.  It is  flexibility for people  who use others component for generate icons.  And it will  resolve problem from my [issue](https://github.com/vuetifyjs/vuetify/issues/4340)

asvae commented:
@johnleider hi.

PR is mostly ready. Just need some help with docs. : 3

ekoeryanto commented:
I was hoping it's included in 1.3

peluprvi commented:
@asvae I created a PR to resolve the merge conflicts. Please check https://github.com/asvae/vuetify/pull/1

To a better workflow, usually branches are used and avoid work directly in the dev branch.
In that case, I created a https://github.com/peluprvi/vuetify/tree/feat/svg-icon-support

asvae commented:
@peluprvi thanks!

I'll check that tomorrow morning (quite late already at my place).

peluprvi commented:
@asvae Also, you forgot to add the `<font-awesome-icon />` component import (or the package) in your playground.
It would be awesome if you could include a `@mdi/js` use case scenario too (https://www.npmjs.com/package/@mdi/js or https://github.com/Templarian/MaterialDesign-JS)

asvae commented:
@peluprvi I think we can even add preset for mdijs (same as I did for FA SVG's).

I'll plan this to holidays.

asvae commented:
Tested and removed conflicts.
Up for merge.

johnleider commented:
Went over this tonight with @KaelWD and like where it's at. Could you add documentation about this on the icons page?

asvae commented:
@johnleider 
> Could you add documentation about this on the icons page?
I already did [PR to docs](https://github.com/vuetifyjs/vuetifyjs.com/pull/621/files). Something is missing there?

johnleider commented:
The docs are no longer their own repo, they are part of the vuetify monorepo and exist right here! https://github.com/vuetifyjs/vuetify/tree/master/packages/vuetifyjs.com

asvae commented:
@johnleider 
Docs up. Tested on my side just in case.

asvae commented on packages/vuetifyjs.com/src/pages/framework/IconsPage.vue:
```diff
@@ -88,6 +88,29 @@
         |  iconfont: 'fa'
         | })
 
+      // FA SVG
```
Will be able to finish this on weekend.

nekosaur commented:
I tried to take a look at this just now, but the svg icons did not display correctly just by following the instructions. See https://i.imgur.com/fnExrXB.png

I had to set a default height when rendering svgs (24px like for regular v-icons), and add `fill: currentColor` so that svg icon would apply any color. Or am I missing something?

asvae commented:
I've checked out from `dev`.
Then did `lerna  bootstrap` (didn't find this in manual, just assumed a correct thing to do).
Then `yarn start`.

Here's what I'm shown in browser: https://media.discordapp.net/attachments/357991952622288896/521435188832632833/unknown.png?width=1443&height=570

asvae commented:
@nekosaur I tried to run Playground and everything was ok. And no, you shouldn't have to tweak anything for resizes to work properly.
I updated `details` in PR to contain literally everything I changed. So you can check.

asvae commented:
@johnleider changes up.

jacekkarczmarczyk commented:
Followed the updated details and displays properly for me. I'm only not sure about the big size of the svg icon

![image](https://user-images.githubusercontent.com/15625235/49707792-851a8300-fc5f-11e8-9377-ea9b74241329.png)

Probably we also don't need `&--{SIZE}` styles as we're setting the `font-weight` in the component

How about adding this?
```diff
diff --git a/packages/vuetify/src/stylus/components/_icons.styl b/packages/vuetify/src/stylus/components/_icons.styl
index 69759ab07..4756abfc8 100644
--- a/packages/vuetify/src/stylus/components/_icons.styl
+++ b/packages/vuetify/src/stylus/components/_icons.styl
@@ -28,14 +28,8 @@ theme(v-icon, "v-icon")
   &.v-icon.v-icon--link
     cursor: pointer

-  &--medium
-    font-size: 2rem
-
-  &--large
-    font-size: 2.5rem
-
-  &--x-large
-    font-size: 3rem
-
   &--disabled
     pointer-events: none
+
+svg.v-icon
+  height: 24px

```
?

![image](https://user-images.githubusercontent.com/15625235/49707852-ba26d580-fc5f-11e8-946c-a88350468d87.png)

nekosaur commented:
@jacekkarczmarczyk That is how it looked for me too, and it's not correct (when comparing to image in PR). Both sizing and color is missing.

jacekkarczmarczyk commented:
Yup, I didn't notice that screenshot in the description, indeed they are missing color too

asvae commented:
So I checked IconSvg and figured I still forgot something.
Which is 
```css
.svg-icon {
  fill: currentColor;
  height: 1.5rem;
}
```

Sorry for being confusing.

asvae commented:
Changes up.

To surmise:
* Removed unused (`&--medium` etc) styles.
* Added some info to manual on how user have to make his SVG icons so they inherit our style and size.
* `disabled` icon went `opacity: .6` (0.6 was already used in a couple of places so I took it).
* Added missing styles to `details` > SvgIcon.vue.

asvae commented:
@blowsie,

> Does this cater for nuxt-fontawesome
If you mean FA SVG preset - then likely yes. It just uses globally declared component (which you can do by any means).

> Also how does this impact `<VToolbarSideIcon>`
And why would it?
Tested just in case: [image](https://user-images.githubusercontent.com/5394573/49954301-3562e280-ff11-11e8-8c33-3dcce2525661.png).

blowsie commented:
@asvae 

I'm testing in 1.4 beta.

Im using nuxt-fontawesome like so...
```
 [
    'nuxt-fontawesome',
    {
      imports: [
        {
          set: '@fortawesome/pro-regular-svg-icons',
          icons: ['faVial', 'faBars', 'faAt', 'faEnvelope', 'faCircle']
        }
      ]
    }
  ],
```

I added the following to my vuetify config
```
  iconfont: 'faSvg',
```

Result
![image](https://user-images.githubusercontent.com/308572/50280843-7ee98b00-0445-11e9-97ab-c8b7eea3b4a7.png)

blowsie commented:
Additionally

There is a significant difference in size.

**Material**
![image](https://user-images.githubusercontent.com/308572/50281503-8e69d380-0447-11e9-9029-7fe08288c9be.png)

**faSvg**

![image](https://user-images.githubusercontent.com/308572/50281534-a2153a00-0447-11e9-8681-6e9503751203.png)

asvae commented:
@blowsie 
First case, if I understood correctly, is that you're able to use `solid` `fa svg` icons (and they work fine) but not `regular` or `light`?
I can address that on holidays.

> There is a significant difference in size
Does this create any problems?

MD handles spacing and sizing differently from FA. MD keeps some space, while FA doesn't.
I don't think it's a good idea to attempt tweaking spacing on our side.

Also, this has nothing to do with FaSvg. FA font icons are also bigger.

![image](https://user-images.githubusercontent.com/5394573/50294815-c6910680-0487-11e9-8d1f-ac785d507516.png)

johnleider commented:
We kindly ask users to not comment on closed/resolved issues. If you believe that this issue has not been correctly resolved, please create a new [issue](https://issues.vuetifyjs.com) showing the regression or reach out to us in our [Discord community](https://community.vuetifyjs.com).

Thank you for your contribution and interest in improving Vuetify.
